File: Blind75/imp interview/invert binary tree.py
# link: https://leetcode.com/problems/invert-binary-tree/submissions/918539000/

'''
technique: Depth frist search
in recrusive way
root remains same
its children is swaped
then next set of tree is swaped
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None

        # swap children
        temp = root.left
        root.left = root.right
        root.right = temp

        self.invertTree(root.left)
        self.invertTree(root.right)

        return root

File: grind75/invert binary tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if (!root) return nullptr;
        
        swap_nodes(root->left, root->right);
        invertTree(root->left);
        invertTree(root->right);
        return root;
        
    }
    void swap_nodes(TreeNode*& node1, TreeNode*& node2) {
        TreeNode* temp = node1;
        node1 = node2;
        node2 = temp;  
    }
};